File: src/main/digestsocial.ts
/**
 * Co dělaly sociální sítě — podklad pro přehled.
 *
 * Aplikace ví o příspěvcích na Instagramu: kdy vyšly, kolik mají lajků
 * a komentářů a na které trhy se rozeslaly. To samo o sobě není tržba, ale
 * je to **jediná věc, kterou e-shop v tom období dělal navenek**, takže když
 * se objednávky hnou, je to první místo, kam se člověk podívá.
 *
 * ## Co se dá spočítat a co ne
 *
 * Spočítat jde, jestli **ve dnech s příspěvkem chodilo víc objednávek** než
 * ve dnech bez něj. Je to korelace, ne důkaz — příspěvek se často pouští
 * právě ve chvíli, kdy je co nabídnout — a tak se to i píše. Je to ale
 * dost na to, aby se dalo poznat, že se za poslední měsíc nepostovalo vůbec
 * a objednávky mezitím spadly.
 *
 * **Zhlédnutí a dosah aplikace nemá.** Instagram je vydává jen přes rozhraní
 * `insights` u vlastního účtu a to se zatím nestahuje; lajky a komentáře ano,
 * ty chodí spolu s příspěvkem. Kdyby se dosah někdy dotahoval, stačí ho
 * přidat sem — zbytek přehledu se měnit nemusí.
 */
import { getDb } from './db';

export interface SocialPost {
  at: string;
  caption: string;
  /** Lajky a komentáře jsou vždycky z Instagramu — Facebook metriky nedává */
  likes: number;
  comments: number;
  permalink: string;
  /** Na kolik trhů se příspěvek rozeslal */
  markets: number;
  /** Které trhy to byly — „CZ, SK, EN" */
  marketLabels?: string[];
  /** „IG" nebo „IG + FB" podle toho, kam se sdílelo */
  channels?: string;
  /**
   * Byl za příspěvkem placený dosah?
   *
   * `null` znamená **nevíme** — starší napojení propagaci nehlásí a „nevíme"
   * se nesmí tvářit jako „ne". Bez tohohle se úspěch koupeného dosahu čte
   * jako úspěch příspěvku a starší propagovaný kus přebije všechno ostatní.
   */
  boosted?: boolean | null;
  /** O kolik % víc objednávek chodilo kolem vydání; null = nedá se spočítat */
  lift?: number | null;
  /** Proč je tenhle příspěvek v seznamu — počítáno z čísel vedle */
  why?: string;
}

export interface SocialView {
  /** Kolik příspěvků vyšlo v okně */
  posts: number;
  likes: number;
  comments: number;
  /** Nejúspěšnější příspěvek okna podle lajků a komentářů */
  best: SocialPost | null;
  /** Kolik dní v okně mělo příspěvek */
  daysWithPost: number;
  /** Průměr objednávek ve dnech s příspěvkem a bez něj */
  ordersWithPost: number;
  ordersWithout: number;
  /** Kolik příspěvků bylo v předchozím okně — na srovnání aktivity */
  prevPosts: number;
  /**
   * Nejúspěšnější příspěvky **z poslední doby** (půl roku).
   *
   * Hlavní pohled je na to, co funguje teď — co zabralo před dvěma lety, je
   * zajímavé u sezóny, ne u rozhodnutí, co postnout příští týden.
   */
  bestEver: SocialPost[];
  /** Doplněk: co fungovalo dávno, ale stojí za připomenutí */
  bestOlder: SocialPost[];
  /** Čerstvé neplacené příspěvky, kterým by rozpočet mohl pomoct */
  candidates: SocialPost[];
  /** Hlásí Instagram u tohohle napojení propagaci? Bez toho se nedá odlišit placené */
  boostKnown: boolean;
}

/**
 * Přehled sítí za okno.
 *
 * `days` je denní řada z přehledu (den a počet objednávek), aby se dvakrát
 * nepočítalo totéž a aby srovnání sedělo přesně na dny, které jsou v grafu.
 */
export function socialView(
  days: { day: string; orders: number }[], windowDays = 30
): SocialView | null {
  const d = getDb();
  const empty: SocialView = {
    posts: 0, likes: 0, comments: 0, best: null, daysWithPost: 0,
    ordersWithPost: 0, ordersWithout: 0, prevPosts: 0, bestEver: [],
    bestOlder: [], candidates: [], boostKnown: false
  };

  let rows: any[] = [];
  try {
    const from = days[0]?.day ?? '';
    const prevFrom = new Date(new Date(`${from}T12:00:00`).getTime() - windowDays * 86_400_000);
    const prevKey = Number.isNaN(prevFrom.getTime()) ? from : prevFrom.toISOString().slice(0, 10);
    rows = d.prepare(
      `SELECT posted_at, caption, like_count, comment_count, permalink, ig_media_id
         FROM ig_source_posts WHERE substr(posted_at, 1, 10) >= ? ORDER BY posted_at DESC LIMIT 200`
    ).all(prevKey) as any[];
  } catch {
    // Instagram v téhle instalaci vůbec není — přehled se kvůli tomu nemění
    return null;
  }
  if (rows.length === 0) return empty;

  const inWindow = new Set(days.map(one => one.day));
  const fromDay = days[0]?.day ?? '';

  const published = (mediaId: string): number => {
    try {
      return Number((d.prepare(
        'SELECT COUNT(*) AS n FROM ig_published WHERE source_media_id = ?'
      ).get(String(mediaId)) as any)?.n ?? 0);
    } catch {
      return 0;
    }
  };

  const postDays = new Set<string>();
  let posts = 0;
  let likes = 0;
  let comments = 0;
  let prevPosts = 0;
  let best: SocialPost | null = null;

  for (const row of rows) {
    const day = String(row.posted_at ?? '').slice(0, 10);
    if (!day) continue;
    if (!inWindow.has(day)) {
      if (day < fromDay) prevPosts++;
      continue;
    }
    posts++;
    postDays.add(day);
    const like = Number(row.like_count ?? 0);
    const comment = Number(row.comment_count ?? 0);
    likes += like;
    comments += comment;
    const one: SocialPost = {
      at: String(row.posted_at ?? ''),
      caption: String(row.caption ?? '').replace(/\s+/g, ' ').trim().slice(0, 120),
      likes: like,
      comments: comment,
      permalink: String(row.permalink ?? ''),
      markets: published(row.ig_media_id)
    };
    // Komentář stojí víc práce než lajk, tak i víc váží
    if (!best || like + comment * 3 > best.likes + best.comments * 3) best = one;
  }

  /*
   * Objednávky ve dnech s příspěvkem a bez něj. Den se počítá celý —
   * příspěvek vyšlý večer se do něj promítne jen zčásti, ale rozlišovat
   * hodiny by u dvou desítek příspěvků nic nepřineslo.
   */
  const withPost = days.filter(one => postDays.has(one.day));
  const without = days.filter(one => !postDays.has(one.day));
  const avg = (list: { orders: number }[]) =>
    list.length ? Math.round((list.reduce((sum, one) => sum + one.orders, 0) / list.length) * 10) / 10 : 0;

  return {
    posts,
    likes,
    comments,
    best,
    daysWithPost: postDays.size,
    ordersWithPost: avg(withPost),
    ordersWithout: avg(without),
    prevPosts,
    /*
     * Dva pohledy zvlášť. Hlavní je poslední půlrok — podle něj se
     * rozhoduje, co postnout teď. Starší úspěchy se přidávají jako
     * připomenutí, ne jako měřítko: co fungovalo před dvěma lety, mohlo
     * mít zaplacený dosah nebo docela jinou nabídku.
     */
    bestEver: bestPosts({ limit: 3, sinceDays: 180 }),
    bestOlder: bestPosts({ limit: 2, beforeDays: 180 }),
    candidates: boostCandidates(days),
    boostKnown: knowsBoost()
  };
}

/* ---------- dlouhodobě ---------- */

/**
 * Kudy příspěvek vyšel.
 *
 * Aplikace publikuje na Instagram a volitelně sdílí na Facebook — a to je
 * jediné, co se o Facebooku dá z databáze zjistit. **Lajky a komentáře jsou
 * vždycky z Instagramu**, protože metriky Facebooku se nikam neukládají;
 * říká se to proto rovnou, ať se čísla nepřipisují oběma sítím.
 */
function channelsOf(sourceMediaId: string): string {
  try {
    const fb = Number((getDb().prepare(
      `SELECT COUNT(*) AS n FROM ig_jobs j
        WHERE j.fb_post_id IS NOT NULL AND j.fb_post_id != ''
          AND j.ig_media_id IN (SELECT ig_media_id FROM ig_published WHERE source_media_id = ?)`
    ).get(String(sourceMediaId)) as any)?.n ?? 0);
    return fb > 0 ? 'IG + FB' : 'IG';
  } catch {
    return 'IG';
  }
}

/** Na které trhy příspěvek šel — jazyky, ne jen počet */
function marketsOf(sourceMediaId: string): string[] {
  try {
    const rows = getDb().prepare(
      'SELECT lang FROM ig_published WHERE source_media_id = ? ORDER BY lang'
    ).all(String(sourceMediaId)) as any[];
    return rows.map(one => String(one.lang ?? '').toUpperCase()).filter(Boolean);
  } catch {
    return [];
  }
}

/**
 * Nejúspěšnější příspěvky za celou historii, případně jen z určitých měsíců.
 *
 * `months` (0 = leden) se hodí u sezóny: „co fungovalo loni v listopadu
 * a prosinci" je pro chystanou kampaň lepší podklad než to, co se povedlo
 * minulý týden. Řadí se podle lajků a komentářů, kde komentář váží víc —
 * napsat ho dá víc práce než klepnout na srdíčko.
 */
export function bestPosts(
  options: { months?: number[]; limit?: number; sinceDays?: number; beforeDays?: number } = {}
): SocialPost[] {
  const limit = options.limit ?? 3;
  let rows: any[] = [];
  try {
    rows = getDb().prepare(
      `SELECT posted_at, caption, like_count, comment_count, permalink, ig_media_id, boosted
         FROM ig_source_posts WHERE posted_at != '' ORDER BY posted_at DESC LIMIT 2000`
    ).all() as any[];
  } catch {
    return [];
  }

  const wanted = options.months;
  const now = Date.now();
  const picked = rows.filter(row => {
    if (wanted?.length) {
      const month = Number(String(row.posted_at ?? '').slice(5, 7)) - 1;
      if (!wanted.includes(month)) return false;
    }
    /*
     * Stáří. Hlavní seznam se dívá na poslední půlrok — podle něj se
     * rozhoduje, co postnout teď; starší se ukazují zvlášť jako připomenutí.
     */
    const when = new Date(String(row.posted_at ?? '')).getTime();
    if (!Number.isFinite(when)) return !options.sinceDays;
    const age = (now - when) / 86_400_000;
    if (options.sinceDays != null && age > options.sinceDays) return false;
    if (options.beforeDays != null && age <= options.beforeDays) return false;
    return true;
  });

  return picked
    .sort((a, b) => score(b) - score(a))
    .slice(0, limit)
    .map(row => toPost(row));
}

/**
 * Ví se u téhle instalace, které příspěvky byly propagované?
 *
 * Stačí jeden příspěvek s vyplněnou hodnotou — pak Instagram propagaci
 * hlásí a dá se rozlišovat. Když ne, řekne se to nahlas: bez toho by
 * „neplacený úspěch" byl jen dohad.
 */
export function knowsBoost(): boolean {
  try {
    return Number((getDb().prepare(
      'SELECT COUNT(*) AS n FROM ig_source_posts WHERE boosted IS NOT NULL'
    ).get() as any)?.n ?? 0) > 0;
  } catch {
    return false;
  }
}

/* ---------- co má smysl propagovat ---------- */

/**
 * Příspěvky, které si říkají o rozpočet.
 *
 * Úspěch placeného příspěvku není zásluha příspěvku — je koupený. Když se
 * pak žebříček řadí jen podle lajků, starší propagovaný příspěvek přebije
 * všechno ostatní a vypadá to, že „tohle funguje", i když to jen mělo
 * zaplacený dosah. Proto se hledá opak: **čerstvý příspěvek, který si vede
 * nadprůměrně bez placení** — u toho má přidání rozpočtu smysl, protože
 * začíná z něčeho, co lidi zabralo samo.
 *
 * Měřítko je medián nedávných neplacených příspěvků, ne průměr: jeden
 * virál by průměr vytáhl tak, že by pak neprošlo nic.
 *
 * `days` je denní řada objednávek z přehledu — u každého kandidáta se
 * přidá, kolik objednávek chodilo v den vydání a dva dny po něm proti
 * běžnému dni. Je to **souvislost, ne důkaz**, a tak se to i píše.
 */
export function boostCandidates(
  days: { day: string; orders: number }[], limit = 3
): SocialPost[] {
  let rows: any[] = [];
  try {
    rows = getDb().prepare(
      `SELECT posted_at, caption, like_count, comment_count, permalink, ig_media_id, boosted
         FROM ig_source_posts WHERE posted_at != '' ORDER BY posted_at DESC LIMIT 200`
    ).all() as any[];
  } catch {
    return [];
  }
  if (rows.length === 0) return [];

  const now = Date.now();
  const fresh = rows.filter(row => {
    const when = new Date(String(row.posted_at ?? '')).getTime();
    return Number.isFinite(when) && now - when <= 60 * 86_400_000;
  });
  if (fresh.length === 0) return [];

  // Měřítko: medián neplacených z posledního půlroku
  const organic = rows
    .filter(row => row.boosted !== 1)
    .map(row => score(row))
    .sort((a, b) => a - b);
  if (organic.length === 0) return [];
  const median = organic[Math.floor(organic.length / 2)] || 1;

  const daily = new Map(days.map(one => [one.day, one.orders]));
  const average = days.length
    ? days.reduce((sum, one) => sum + one.orders, 0) / days.length
    : 0;

  return fresh
    .filter(row => row.boosted !== 1 && score(row) >= median * 1.3)
    .sort((a, b) => score(b) - score(a))
    .slice(0, limit)
    .map(row => {
      const post = toPost(row);
      /*
       * Objednávky kolem vydání. Den vydání a dva dny po něm — déle už se
       * to míchá s čímkoli jiným, co se ten týden dělo.
       */
      const day = String(row.posted_at ?? '').slice(0, 10);
      const around: number[] = [];
      for (let ahead = 0; ahead <= 2; ahead++) {
        const key = new Date(new Date(`${day}T12:00:00`).getTime() + ahead * 86_400_000)
          .toISOString().slice(0, 10);
        const found = daily.get(key);
        if (found != null) around.push(found);
      }
      const mine = around.length
        ? around.reduce((sum, one) => sum + one, 0) / around.length
        : 0;
      const lift = average > 0 && around.length
        ? Math.round(((mine - average) / average) * 100)
        : null;

      const times = Math.round((score(row) / Math.max(1, median)) * 10) / 10;
      post.lift = lift;
      post.why = `Zaujal ${times}× víc než běžný neplacený příspěvek`
        + (post.boosted === false ? ' a rozpočet za ním nestál' : '')
        + (lift != null
          ? `; v den vydání a dva dny po něm chodilo ${lift > 0 ? `o ${lift} % víc` : lift < 0 ? `o ${-lift} % míň` : 'stejně'} objednávek než obvykle (souvislost, ne důkaz)`
          : '')
        + '.';
      return post;
    });
}

/** Lajk je klepnutí, komentář práce — proto váží víc */
function score(row: any): number {
  return Number(row.like_count ?? 0) + Number(row.comment_count ?? 0) * 3;
}

/** Řádek z databáze na příspěvek i s trhy, kanály a propagací */
function toPost(row: any): SocialPost {
  const mediaId = String(row.ig_media_id ?? '');
  const markets = marketsOf(mediaId);
  return {
    at: String(row.posted_at ?? ''),
    caption: String(row.caption ?? '').replace(/\s+/g, ' ').trim().slice(0, 120),
    likes: Number(row.like_count ?? 0),
    comments: Number(row.comment_count ?? 0),
    permalink: String(row.permalink ?? ''),
    markets: markets.length,
    marketLabels: markets,
    channels: channelsOf(mediaId),
    // `null` = Instagram propagaci u tohohle napojení nehlásí
    boosted: row.boosted == null ? null : row.boosted === 1
  };
}

File: ios/QuentinoApp/Services/DigestSocial.swift
import Foundation

/**
 Co dělaly sociální sítě — podklad pro AI přehled.

 Sesterský modul k `src/main/digestsocial.ts`. Aplikace ví o příspěvcích na
 Instagramu: kdy vyšly, kolik mají lajků a komentářů a na kolik trhů se
 rozeslaly. Spočítat jde, jestli **ve dnech s příspěvkem chodilo víc
 objednávek** — je to souvislost, ne důkaz, a tak se to i píše.

 Zhlédnutí a dosah aplikace nemá: Instagram je vydává jen přes rozhraní
 `insights` a to se zatím nestahuje.
 */
enum DigestSocial {
    static func view(days: [[String: Any]], windowDays: Int = 30) -> Any {
        let fromDay = days.first?["day"] as? String ?? ""
        guard !fromDay.isEmpty else { return NSNull() }

        // O kolik dní zpátky se ptát, aby bylo i předchozí období na srovnání
        let prevKey: String = {
            guard let from = Formats.date("\(fromDay) 12:00:00") ?? isoDay(fromDay) else { return fromDay }
            let back = from.addingTimeInterval(-Double(windowDays) * 86_400)
            return String(Formats.iso(back).prefix(10))
        }()

        guard let rows = try? SQLite.shared.query(
            "SELECT posted_at, caption, like_count, comment_count, permalink, ig_media_id "
            + "FROM ig_source_posts WHERE substr(posted_at, 1, 10) >= ? ORDER BY posted_at DESC LIMIT 200",
            [.text(prevKey)]
        ) else {
            // Instagram v téhle instalaci vůbec není — přehled se tím nemění
            return NSNull()
        }

        var inWindow = Set<String>()
        for one in days { if let day = one["day"] as? String { inWindow.insert(day) } }

        var postDays = Set<String>()
        var posts = 0
        var likes = 0
        var comments = 0
        var prevPosts = 0
        var best: [String: Any]?
        var bestScore = -1

        for row in rows {
            let day = String((row["posted_at"] as? String ?? "").prefix(10))
            if day.isEmpty { continue }
            if !inWindow.contains(day) {
                if day < fromDay { prevPosts += 1 }
                continue
            }
            posts += 1
            postDays.insert(day)
            let like = row["like_count"] as? Int ?? 0
            let comment = row["comment_count"] as? Int ?? 0
            likes += like
            comments += comment

            // Komentář stojí víc práce než lajk, tak i víc váží
            let score = like + comment * 3
            if score > bestScore {
                bestScore = score
                var one: [String: Any] = [:]
                one["at"] = row["posted_at"] as? String ?? ""
                let caption = (row["caption"] as? String ?? "")
                    .replacingOccurrences(of: "\\s+", with: " ", options: .regularExpression)
                    .trimmingCharacters(in: .whitespacesAndNewlines)
                one["caption"] = String(caption.prefix(120))
                one["likes"] = like
                one["comments"] = comment
                one["permalink"] = row["permalink"] as? String ?? ""
                one["markets"] = markets(row["ig_media_id"] as? String ?? "")
                best = one
            }
        }

        let withPost = days.filter { postDays.contains($0["day"] as? String ?? "") }
        let without = days.filter { !postDays.contains($0["day"] as? String ?? "") }

        var out: [String: Any] = [:]
        out["posts"] = posts
        out["likes"] = likes
        out["comments"] = comments
        out["best"] = best ?? NSNull()
        out["daysWithPost"] = postDays.count
        out["ordersWithPost"] = average(withPost)
        out["ordersWithout"] = average(without)
        out["prevPosts"] = prevPosts
        /*
         Dva pohledy zvlášť. Hlavní je poslední půlrok — podle něj se
         rozhoduje, co postnout teď. Starší úspěchy se přidávají jako
         připomenutí, ne jako měřítko: mohly mít zaplacený dosah.
         */
        out["bestEver"] = bestPosts(limit: 3, sinceDays: 180)
        out["bestOlder"] = bestPosts(limit: 2, beforeDays: 180)
        out["candidates"] = boostCandidates(days)
        out["boostKnown"] = knowsBoost()
        return out
    }

    /**
     Kudy příspěvek vyšel.

     Aplikace publikuje na Instagram a volitelně sdílí na Facebook — a to je
     jediné, co se o Facebooku dá z databáze zjistit. **Lajky a komentáře
     jsou vždycky z Instagramu**, protože metriky Facebooku se nikam
     neukládají; říká se to proto rovnou.
     */
    private static func channels(_ mediaId: String) -> String {
        guard !mediaId.isEmpty else { return "IG" }
        let count = ((try? SQLite.shared.query(
            "SELECT COUNT(*) AS n FROM ig_jobs j WHERE j.fb_post_id IS NOT NULL AND j.fb_post_id != '' "
            + "AND j.ig_media_id IN (SELECT ig_media_id FROM ig_published WHERE source_media_id = ?)",
            [.text(mediaId)]
        ))?.first?["n"] as? Int) ?? 0
        return count > 0 ? "IG + FB" : "IG"
    }

    /// Na které trhy příspěvek šel — jazyky, ne jen počet
    private static func marketLabels(_ mediaId: String) -> [String] {
        guard !mediaId.isEmpty else { return [] }
        let rows = (try? SQLite.shared.query(
            "SELECT lang FROM ig_published WHERE source_media_id = ? ORDER BY lang", [.text(mediaId)])) ?? []
        return rows.compactMap { ($0["lang"] as? String)?.uppercased() }.filter { !$0.isEmpty }
    }

    /**
     Nejúspěšnější příspěvky za celou historii, případně jen z určitých měsíců.

     `months` (0 = leden) se hodí u sezóny: „co fungovalo loni v listopadu
     a prosinci" je pro chystanou kampaň lepší podklad než minulý týden.
     */
    static func bestPosts(
        months: [Int] = [], limit: Int = 3, sinceDays: Int? = nil, beforeDays: Int? = nil
    ) -> [[String: Any]] {
        let rows = sourceRows()

        let wanted = Set(months)
        let now = Date().timeIntervalSince1970
        var picked: [(row: [String: Any], score: Int)] = []
        for row in rows {
            if !wanted.isEmpty {
                let month = Int(String((row["posted_at"] as? String ?? "").dropFirst(5).prefix(2))) ?? 0
                if !wanted.contains(month - 1) { continue }
            }
            /*
             Stáří. Hlavní seznam se dívá na poslední půlrok — podle něj se
             rozhoduje, co postnout teď; starší se ukazují zvlášť jako
             připomenutí, protože mohly mít zaplacený dosah.
             */
            if sinceDays != nil || beforeDays != nil {
                guard let when = Formats.date(row["posted_at"] as? String ?? "") else {
                    if sinceDays != nil { continue }
                    picked.append((row, score(row)))
                    continue
                }
                let age = (now - when.timeIntervalSince1970) / 86_400
                if let sinceDays, age > Double(sinceDays) { continue }
                if let beforeDays, age <= Double(beforeDays) { continue }
            }
            picked.append((row, score(row)))
        }

        return picked.sorted { $0.score > $1.score }.prefix(limit).map { found in post(found.row) }
    }

    /// Lajk je klepnutí, komentář práce — proto váží víc
    private static func score(_ row: [String: Any]) -> Int {
        (row["like_count"] as? Int ?? 0) + (row["comment_count"] as? Int ?? 0) * 3
    }

    /// Zdrojové příspěvky i s propagací; starší databáze sloupec nemá
    private static func sourceRows() -> [[String: Any]] {
        let withBoost = (try? SQLite.shared.query(
            "SELECT posted_at, caption, like_count, comment_count, permalink, ig_media_id, boosted "
            + "FROM ig_source_posts WHERE posted_at != '' ORDER BY posted_at DESC LIMIT 2000")) ?? []
        if !withBoost.isEmpty { return withBoost }
        return (try? SQLite.shared.query(
            "SELECT posted_at, caption, like_count, comment_count, permalink, ig_media_id "
            + "FROM ig_source_posts WHERE posted_at != '' ORDER BY posted_at DESC LIMIT 2000")) ?? []
    }

    /// Řádek z databáze na příspěvek i s trhy, kanály a propagací
    private static func post(_ row: [String: Any]) -> [String: Any] {
        let mediaId = row["ig_media_id"] as? String ?? ""
        let caption = (row["caption"] as? String ?? "")
            .replacingOccurrences(of: "\\s+", with: " ", options: .regularExpression)
            .trimmingCharacters(in: .whitespacesAndNewlines)
        let labels = marketLabels(mediaId)

        var one: [String: Any] = [:]
        one["at"] = row["posted_at"] as? String ?? ""
        one["caption"] = String(caption.prefix(120))
        one["likes"] = row["like_count"] as? Int ?? 0
        one["comments"] = row["comment_count"] as? Int ?? 0
        one["permalink"] = row["permalink"] as? String ?? ""
        one["markets"] = labels.count
        one["marketLabels"] = labels
        one["channels"] = channels(mediaId)
        // Prázdná hodnota znamená nevíme — a nevíme se nesmí tvářit jako ne
        if let boosted = row["boosted"] as? Int {
            one["boosted"] = boosted == 1
        } else {
            one["boosted"] = NSNull()
        }
        return one
    }

    /**
     Příspěvky, které si říkají o rozpočet.

     Úspěch placeného příspěvku není zásluha příspěvku — je koupený. Hledá se
     proto opak: čerstvý příspěvek, který si vede nadprůměrně bez placení.
     Měřítko je medián neplacených, ne průměr — jeden virál by průměr vytáhl
     tak, že by pak neprošlo nic.
     */
    static func boostCandidates(_ days: [[String: Any]], limit: Int = 3) -> [[String: Any]] {
        let rows = sourceRows()
        guard !rows.isEmpty else { return [] }

        let now = Date().timeIntervalSince1970
        let fresh = rows.filter { row in
            guard let when = Formats.date(row["posted_at"] as? String ?? "") else { return false }
            return now - when.timeIntervalSince1970 <= 60 * 86_400
        }
        guard !fresh.isEmpty else { return [] }

        let organic = rows.filter { ($0["boosted"] as? Int) != 1 }.map { score($0) }.sorted()
        guard !organic.isEmpty else { return [] }
        let median = max(1, organic[organic.count / 2])

        var daily: [String: Int] = [:]
        for day in days { daily[day["day"] as? String ?? ""] = day["orders"] as? Int ?? 0 }
        let average = days.isEmpty
            ? 0.0
            : Double(days.reduce(0) { $0 + ($1["orders"] as? Int ?? 0) }) / Double(days.count)

        let picked = fresh
            .filter { ($0["boosted"] as? Int) != 1 && Double(score($0)) >= Double(median) * 1.3 }
            .sorted { score($0) > score($1) }
            .prefix(limit)

        return picked.map { row in
            var one = post(row)
            /*
             Objednávky kolem vydání — den vydání a dva dny po něm. Déle už
             se to míchá s čímkoli jiným, co se ten týden dělo. Je to
             souvislost, ne důkaz, a tak se to i píše.
             */
            let day = String((row["posted_at"] as? String ?? "").prefix(10))
            var around: [Int] = []
            if let start = Formats.date(day + "T12:00:00Z") {
                for ahead in 0...2 {
                    let key = String(Formats.iso(start.addingTimeInterval(Double(ahead) * 86_400)).prefix(10))
                    if let value = daily[key] { around.append(value) }
                }
            }
            let mine = around.isEmpty ? 0 : Double(around.reduce(0, +)) / Double(around.count)
            let lift: Int? = (average > 0 && !around.isEmpty)
                ? Int((((mine - average) / average) * 100).rounded())
                : nil
            let times = (Double(score(row)) / Double(median) * 10).rounded() / 10

            var why = "Zaujal \(times)× víc než běžný neplacený příspěvek"
            if (row["boosted"] as? Int) == 0 { why += " a rozpočet za ním nestál" }
            if let lift {
                let how = lift > 0 ? "o \(lift) % víc" : (lift < 0 ? "o \(-lift) % míň" : "stejně")
                why += "; v den vydání a dva dny po něm chodilo \(how) objednávek než obvykle"
                    + " (souvislost, ne důkaz)"
            }
            one["why"] = why + "."
            one["lift"] = lift ?? NSNull()
            return one
        }
    }

    /// Ví se u téhle instalace, které příspěvky byly propagované?
    static func knowsBoost() -> Bool {
        let count = ((try? SQLite.shared.query(
            "SELECT COUNT(*) AS n FROM ig_source_posts WHERE boosted IS NOT NULL"
        ))?.first?["n"] as? Int) ?? 0
        return count > 0
    }

    private static func average(_ days: [[String: Any]]) -> Double {
        guard !days.isEmpty else { return 0 }
        let sum = days.reduce(0) { $0 + ($1["orders"] as? Int ?? 0) }
        return (Double(sum) / Double(days.count) * 10).rounded() / 10
    }

    private static func markets(_ mediaId: String) -> Int {
        guard !mediaId.isEmpty else { return 0 }
        return ((try? SQLite.shared.query(
            "SELECT COUNT(*) AS n FROM ig_published WHERE source_media_id = ?", [.text(mediaId)]
        ))?.first?["n"] as? Int) ?? 0
    }

    /// `YYYY-MM-DD` na datum — na poledne, ať časové pásmo nepřehodí den
    private static func isoDay(_ key: String) -> Date? {
        let parts = key.split(separator: "-").compactMap { Int($0) }
        guard parts.count == 3 else { return nil }
        var components = DateComponents()
        components.year = parts[0]
        components.month = parts[1]
        components.day = parts[2]
        components.hour = 12
        return Calendar.current.date(from: components)
    }
}
